File: activity/20200904/scau-flyingbird/CoinChange.go
package main

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	for i := 0; i <= amount; i++ {
		dp[i] = amount + 1
	}
	dp[0] = 0
	for i := 1; i < len(coins); i++ {
		for j := 1; j <= amount; j++ {
			if j >= coins[i-1] {
				dp[j] = Min(dp[j], dp[j-coins[i-1]]+1)
			}
		}
	}

	if dp[amount] == amount+1 {
		return -1
	} else {
		return dp[amount]
	}
}

func Min(a, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}

File: activity/20200904/RabbitLee/Coin_Change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [-1 for i in range(amount + 1)]
        dp[0] = 0
        for i in range(1, amount + 1):
            res = i + 1
            for coin in coins:
                if i >= coin and dp[i - coin] >= 0:
                    res = min(res, 1 + dp[i - coin])
            if res <= i:
                dp[i] = res
        return dp[amount]

